File: cached-20200215/python/node/Node.py
#! -*- encoding=utf-8 -*-

import os

class Node:
	def __init__(self,key,value):
		self.key = key
		self.value = value
		self.prev = None
		self.next =None

	def __str__(self):
		# 返回一个对象的描述信
		return '{%d: %d}' % (self.key,self.value)
	def __repr__(self):
		# 类名+object at+内存地址
		return '{%d: %d}' % (self.key,self.value)

class DoubleLinkedList:
	"""一个简单的双向列表"""
	def __init__(self,capacity=0xffff):
		self.capacity = capacity
		self.head = None
		self.tail = None
		self.size = 0
		
	# 从头部添加
	def __add_head(self,node):
		if not self.head:
			self.head = node
			self.tail = node
			self.head.next = None
			self.head.prev = None
		else:
			node.next = self.head
			self.head.prev = node
			self.head = node
			self.head.prev = None
		self.size += 1
		return node

	# 从尾部添加
	def __add_tail(self,node):
		if not self.tail:
			# 判断当前节点是否为空的，如果是空的
			self.tail = node
			self.head = node
			self.tail.next = None
			self.tail.prev = None
		else:
			self.tail.next = node
			node.prev = self.tail
			self.tail = node
			self.tail.next = None
		self.size += 1
		return node
 
	#从尾部删除
	def __del_tail(self):
		if not self.tail:	
			return
		node = self.tail
		if node.prev:
			self.tail = node.prev
			self.tail.next = None
		else:
			self.tail = self.head = None
		self.size -=1
		return node
	
	#从头部删除 
	def __del_head(self):
		# pass
		if not self.head:
			return
		node = self.head
		if node.next:
			self.head = node.next
			self.head.prev = None
		else:
			self.tail = self.head = None
		self.size -=1
		return node

	#删除任意节点
	def __remove(self,node):
		# 如果node=None，默认删除尾部节点
		if not node:
			node = self.tail
		if node == self.tail:
			self.__del_tail()
		elif node == self.head:
			self.__del_head()
		else:
			node.prev.next = node.next
			node.next.prev = node.prev
			self.size -= 1
		return node
	
	# 弹出头部节点
	def pop(self):
		return self.__del_head()

	# 追加节点
	def append(self,node):
		return self.__add_tail(node)

	# 头部添加节点
	def append_front(self,node):
		return self.__add_head(node)

	#删除节点
	def remove(self,node=None):
		return self.__remove(node)

	def print(self):   
		p = self.head
		line = ''
		# 打印数据类型
		# print(type(p))
		# 打印类成员属性
		# print(p.__dict__)
		 
		while p:
			line += '%s' % p
			p = p.next

			# print(a);
			if p:
				line += '=>'
		print(line)

# 测试双向链表
if __name__ == '__main__':
	l = DoubleLinkedList(10)
	nodes = []
	for i in range(10):
		nodes.append(Node(i,i))
	l.append(nodes[0])
	l.print()
	l.append(nodes[1])
	l.print()
	
	l.pop()
	l.print()
	l.append(nodes[2])
	l.print()

	l.append_front(nodes[3])
	l.print()
	l.append(nodes[4])
	l.print()
	l.remove(nodes[2])
	l.print()
	l.remove()
	l.print()
	print(os.getcwd())

File: cached-20200215/php/node/Node.php
<?php

class Node{
	protected $key = NULL;//键
	protected $value = NULL;//值
	protected $prev = NULL;//前驱
	protected $next = NULL;//后驱
	public function __construct($key,$value){
		$this->key = $key;
		$this->value = $value;	
	}
	public function __toString(){
		return sprintf('%s => %s',$this->key,$this->value);
	}
	/*
	php获取内存地址：
	因为传统PHP是单线程的，从设计思想上来说是不支持获取内存地址的，这样可以使程序员更注重于业务开发，而且获取到地址用处也有限，如果非要获取内存地址可以使用&符号代替。
	 */
}
class DoubleLinkedList extends Node{
	/*一个简单的双向列表*/
	private $head;
	private $tail;
	protected $size = 0;
	private $capacity = 0xffff;

	public function __construct($capacity=0xffff){
		$this->capacity = $capacity;
		$this->head = $this->tail = new Node(NULL,NULL);
	}

	//头部添加
	private function __add_head($node){
		
		if(self::is_empty($this->head->key)){
			$this->head = $node;
			$this->tail = $node;
			$this->head->prev = NULL;
			$this->head->next = NULL;
		}else{
			$node->next = $this->head;
			$this->head->prev = $node;
			$this->head = $node;
			$this->head->prev = NULL;
		}
		$this->size += 1;
		return $node;
	}

	//尾部追加
	private function __add_tail($node){
		if(@self::is_empty($this->tail->key)){
			$this->tail = $node;
			$this->head = $node;
			$this->tail->next = NULL;
			$this->tail->prev = NULL;
		}else{
			$this->tail->next = $node;
			$node->prev = $this->tail;
			$this->tail = $node;
			$this->tail->next = NULL;
		}
		$this->size += 1;
		return $node;
	}

	//尾部删除
	private function __del_tail(){
		if(self::is_empty($this->tail->key)) return;
		$node = $this->tail;
		if(!self::is_empty($node->prev)){
			$this->tail = $node->prev;
			$this->tail->next = NULL;
		}else{
			$this->tail = $this->head = NULL;
		}
		$this->size -= 1;
		return $node;
	}

	//头部删除
	private function __del_head(){
		if(self::is_empty($this->head->key)) return;
		$node = $this->head;
		if(@!self::is_empty($node->next->key)){
			$this->head = $node->next;
			$this->head->prev = NULL;
		}else{
			$this->tail = $this->head = NULL;
		}
		$this->size -= 1;
		return $node;
	}

	//删除任意节点
	private function __remove($node){
		//如果node==NULL删除尾部节点
		if($this->is_empty($node)){
			$node = $this->tail;
		}

		if($node == $this->tail){
			$this->__del_tail();
		}elseif($node == $this->head){
			$this->__del_head();
		}else{
			$node->prev->next = $node->next;
			$node->next->prev = $node->prev;
			$this->size -= 1;
		}
		return $node;
	}

	//弹出头部节点
	public function pop(){
		return $this->__del_head();
	}

	//追加节点
	public function append($node){
		return $this->__add_tail($node);
	}

	//头部添加节点
	public function append_front($node){
		return $this->__add_head($node);
	}
	
	//删除节点
	public function remove($node=NULL){
		return $this->__remove($node);
	}
	/*
	 判断当前节点是否为空,为空返回TRUE,不为空返回FALSE
	*/
	public static function is_empty($key){
		return (isset($key) and !is_null($key) and !empty($key)) ? FALSE : TRUE;
	}

	//输出
	public function console(){
		$p = $this->head;		
		$line = '';
		while(!self::is_empty($p)){
			$line .= sprintf("{%s: %s}",$p->key,$p->value);
			$p = $p->next;
			if(!self::is_empty($p)){
				$line .= '=>';
			}
		}
		print($line.PHP_EOL);
	}
}
//测试双向链表
/*$l = new DoubleLinkedList(10);
$nodes = [];
for($i=1;$i<=10;$i++){
 	array_push($nodes,new Node($i,$i));
}
$l->append($nodes[0]);
$l->console();
$l->append($nodes[1]);
$l->console();
$l->pop();
$l->console();
$l->append($nodes[2]);
$l->console();
$l->append_front($nodes[3]);
$l->console();
$l->append($nodes[4]);
$l->console();
$l->remove($nodes[2]);
$l->console();
$l->remove();
$l->console();
*/